src/Search/SearchServiceProvider.php

```diff
@@ -18,63 +18,91 @@
+            ];
+        });
 
-        $this->registerUserGambits();
+        $this->app->singleton('flarum.simple_search.gambits', function () {
+            return [
+                DiscussionSearcher::class => [
+                    AuthorGambit::class,
+                    CreatedGambit::class,
+                    HiddenGambit::class,
+                    UnreadGambit::class
+                ],
+                UserSearcher::class => [
+                    EmailGambit::class,
+                    GroupGambit::class
+                ]
+            ];
+        });
+
+        $this->app->singleton('flarum.simple_search.search_mutators', function () {
+            return [];
+        });
     }
 
-    public function registerUserGambits()
+    /**
+     * {@inheritdoc}
+     */
+    public function boot()
     {
-        $this->app->when(UserSearcher::class)
-            ->needs(GambitManager::class)
-            ->give(function (Container $app) {
-                $gambits = new GambitManager($app);
-
-                $gambits->setFulltextGambit(UserFulltextGambit::class);
-                $gambits->add(EmailGambit::class);
-                $gambits->add(GroupGambit::class);
+        // The rest of these we can resolve in the when->needs->give callback,
+        // but we need to resolve at least one regardless so we know which
+        // searchers we need to register gambits for.
+        $fullTextGambits = $this->app->make('flarum.simple_search.fulltext_gambits');
 
-                $app->make('events')->dispatch(
-                    new ConfigureUserGambits($gambits)
-                );
+        foreach ($fullTextGambits as $searcher => $fullTextGambitClass) {
+            $this->app
+                ->when($searcher)
+                ->needs(GambitManager::class)
+                ->give(function () use ($searcher, $fullTextGambitClass) {
+                    $gambitManager = new GambitManager();
+                    $gambitManager->setFulltextGambit($this->app->make($fullTextGambitClass));
+                    foreach (Arr::get($this->app->make('flarum.simple_search.gambits'), $searcher, []) as $gambit) {
+                        $gambitManager->add($this->app->make($gambit));
+                    }
 
-                return $gambits;
-            });
-    }
-
-    public function registerDiscussionGambits()
-    {
-        $this->app->when(DiscussionSearcher::class)
-            ->needs(GambitManager::class)
-            ->give(function (Container $app) {
-                $gambits = new GambitManager($app);
+                    if ($searcher === UserSearcher::class) {
+                        $this->app->make('events')->dispatch(
+                            new ConfigureUserGambits($gambitManager)
+                        );
+                    } elseif ($searcher === DiscussionSearcher::class) {
+                        $this->app->make('events')->dispatch(
+                            new ConfigureDiscussionGambits($gambitManager)
```

Comment:
The gambits added by old extensions will be classes and not instances, so they won't work. (same with full text gambits)

Let's do the same here with what you did before, add them to a temporary gambit manager, get all gambits and re-add them to the original gambit manager after making them.

Reply:
Ah, that's why I did that! I was confused in retrospect.